db/lens.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+
+	ds "github.com/ipfs/go-datastore"
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/db/description"
+	"github.com/sourcenetwork/defradb/errors"
+)
+
+func (db *db) setMigration(ctx context.Context, txn datastore.Txn, cfg client.LensConfig) error {
+	dstCols, err := description.GetCollectionsBySchemaVersionID(ctx, txn, cfg.DestinationSchemaVersionID)
+	if err != nil {
+		return err
+	}
+
+	sourceCols, err := description.GetCollectionsBySchemaVersionID(ctx, txn, cfg.SourceSchemaVersionID)
+	if err != nil {
+		return err
+	}
+
+	colSeq, err := db.getSequence(ctx, txn, core.COLLECTION)
+	if err != nil {
+		return err
+	}
+
+	if len(sourceCols) == 0 {
+		// If no collections are found with the given [SourceSchemaVersionID], this migration must be from
+		// a collection/schema version that does not yet exist locally.  We must now create it.
+		colID, err := colSeq.next(ctx, txn)
+		if err != nil {
+			return err
+		}
+
+		desc := client.CollectionDescription{
+			ID:              uint32(colID),
+			RootID:          client.OrphanRootID,
+			SchemaVersionID: cfg.SourceSchemaVersionID,
+		}
+
+		col, err := description.SaveCollection(ctx, txn, desc)
+		if err != nil {
+			return err
+		}
+
+		sourceCols = append(sourceCols, col)
+	}
+
+	for _, sourceCol := range sourceCols {
```

Comment:
:) I do disagree with this, and do see it as a personal styling preference of yours.  I find a more 'modern' `for i, foo` loop to be much more readable and less error prone.  The performance difference can go either way too, and is typically very very small.

I do see `courseCol := &courseCols[i]` as the worst of both worlds lol.  Especially when it is creating a pointer.

Reply:
> a personal styling preference of yours

It's not personal. I think it is everybody's preference to have faster code than slower.

Yes, you are right, "modern" style is a bit more readable, but I don't think it outweighs the benefits of faster code.
If you look at one instance of a loop it seem like very small impact, but this approach is used in lots of places, which means we do a lot of copying. This adds up. 

> I do see courseCol := &courseCols[i] as the worst of both worlds lol. Especially when it is creating a pointer.

How is it worst?

Reply:
> It's not personal. I think it is everybody's preference to have faster code than slower.

Performance is not the ultimate goal of most code.  Typically we trade performance for readability (otherwise we'd all code in assembly).

Your suggestion is not always going to be more performant either. `foo[i]` is not free.

> How is it worst?

It does what `for i, foo` does, but in a more unexpected way, and it uses a pointer.  Depending on the size of `foo` that can be more expensive than copying, and it can result in more GC pressure.

----

In my opinion:

Using `for i {` everywhere is seen by me as a premature optimisation with an unproven and likely negligible impact on our users and a very real negative impact on readability.

And readability issues breed bad and unperformant code.  Prioritising blanket micro-optimisations like this could ultimately result in much worse performance for our users.

And given that we differ, and a very significant part of our profession appears to also differ (else `for i, foo` would not exist in Go and other langs), I see this as a personal preference and is not something that should be imposed on any developer.

Reply:
> Using for i { everywhere

I didn't say everywhere. Some types (like ints) are much cheaper to copy by values. I should be decided on case-by-case basis.
`CollectionDescription` is cheaper to pass around by pointer.

Anyway... didn't want to start a long conversation. It's a suggestion after all.

Reply:
My opinion is that `for i, foo := range bar` improves readability and the extra allocation here has very little impact on performance. Not worth switching to `for i := range bar`.